### ISSUE TITLE
SEV-ES: error: kvm run failed Invalid argument

### ISSUE THREAD

**Issue by Alexandre-Neves** [AMDESE/AMDSEV#58]
I've successfully followed the instructions in the repository.

When running `./launch-qemu.sh -hda myfile.qcow2 -vnc 1 -console serial -sev-es` I get the following, after choosing which OS to boot in the VM in the bootloader:

`...
CpuDxe: 5-Level Paging = 0
MpInitChangeApLoopCallback() done!
SetUefiImageMemoryAttributes - 0x000000007F2E7000 - 0x0000000000008000 (0x0000000000000008)
SetUefiImageMemoryAttributes - 0x000000007F2E1000 - 0x0000000000006000 (0x0000000000000008)
SetUefiImageMemoryAttributes - 0x000000007F2DB000 - 0x0000000000006000 (0x0000000000000008)
SetUefiImageMemoryAttributes - 0x000000007F2D5000 - 0x0000000000006000 (0x0000000000000008)
SetUefiImageMemoryAttributes - 0x000000007F2C6000 - 0x000000000000F000 (0x0000000000000008)
SetUefiImageMemoryAttributes - 0x000000007F2C1000 - 0x0000000000005000 (0x0000000000000008)
SetUefiImageMemoryAttributes - 0x000000007F2BB000 - 0x0000000000006000 (0x0000000000000008)
SetUefiImageMemoryAttributes - 0x000000007F2B7000 - 0x0000000000004000 (0x0000000000000008)
SetUefiImageMemoryAttributes - 0x000000007F2B3000 - 0x0000000000004000 (0x0000000000000008)
error: kvm run failed Invalid argument
EAX=00800f12 EBX=00000000 ECX=004003f1 EDX=2fd3fbff
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=00000000 EFL=00000002 [-------] CPL=0 II=0 A20=1 SMM=0 HLT=0
ES =0000 00000000 00000000 00000000
CS =0000 00000000 00000000 00000000
SS =0000 00000000 00000000 00000000
DS =0000 00000000 00000000 00000000
FS =0000 00000000 00000000 00000000
GS =0000 00000000 00000000 00000000
LDT=0000 00000000 00000000 00000000
TR =0000 00000000 00000000 00000000
GDT=     00000000 00000000
IDT=     00000000 00000000
CR0=00010033 CR2=00000000 CR3=00000000 CR4=00000648
DR0=0000000000000000 DR1=0000000000000000 DR2=0000000000000000 DR3=0000000000000000 
DR6=00000000ffff0ff0 DR7=0000000000000400
EFER=0000000000000800
Code=<dc> 88 bd 0c 68 f1 f7 09 f0 98 01 9c 81 2b c5 4c 74 77 56 40 6f e0 ec dd b8 6e c8 d8 43 6f b2 0f ad 88 69 64 7a 19 9f 34 ee fc 78 4c fa d1 58 26 db 19`

When running `./launch-qemu.sh -hda myfile.qcow2 -vnc 1 -console serial -sev`, I get the same output up until the `error: kvm run failed Invalid argument`, and the VM runs fine. 

Also, by comparing the two commands issued by the script, the only argument differences are `-no-reboot` and the policy value in the `-object` parameter, which is set to `0x5` for the former and omitted for the latter. Even after removing the `-no-reboot` from the `sev-es` run, the error remains.

Just to be clear, in the host OS, `dmesg | grep ASID` yields `[   10.951826] SEV supported: 478 ASIDs [   10.951826] SEV-ES supported: 31 ASIDs`.

What could be causing this?

**Comment by tlendacky** [comment]
It could be the level of the bootloader/grub. I've seen some versions of grub that invoke the guest in 32-bit mode, rather than 64-bit mode, which isn't supported (yet). Try this... boot the VM normally and send a copy of the kernel and initrd to the hypervisor and then try booting the VM again with the -kernel, -initrd and -append arguments (copy/paste the output of the VM /proc/cmdline contents to use) and the -sev-es option.

Joerg is working on a patchset to allow the 32-bit boot path, but I don't think he has finished it yet.

**Comment by Alexandre-Neves** [comment]
I'm not very experienced in working with kernels and VMs, so which file would correspond to the kernel and which one to the initrd?

Also, what is the best way to transfer files from the VM to the host?

Sorry if these questions are not exactly specific to the issue.

**Comment by tlendacky** [comment]
The kernel and initrd will be located under the /boot directory. The kernel will begin with vmlinuz* and the initrd will begin with init*. You can verify the kernel name and initrd name (assuming you're using grub) at the grub boot menu (highlight the entry to boot and press 'e' and scroll down until you see the linux and initrd entries) or by looking at /boot/grub/grub.cfg (although this file is located in various places depending on the OS, it could also be /boot/grub2/grub.cfg or /boot/efi/EFI/**_DISTRO-NAME_**/grub.cfg).

For transferring the files I typically use scp.  Alternatively, you can get the files without booting the VM by mounting the partitions of the qcow2 image and then copying the files directly (google how to mount a qcow2 image).

**Comment by Alexandre-Neves** [comment]
I got it! Thank you so much!

For posterity: I used the `/proc/cmdline` from the original VM's kernel (in my case, 4.19) and changed the value of `BOOT_IMAGE` to match the new kernel. I also changed the `launch-qemu.sh` script to add the `-kernel`, `-append` and `-initrd` options.

Also, mounting a `.qcow2` is incredibly simple and fast: https://gist.github.com/shamil/62935d9b456a6f9877b5

I'm closing this.
